File: python/neutron/nucleus/datalog.py
"""Datalog model — wraps Nucleus DATALOG_* SQL functions."""

from __future__ import annotations

import json
from typing import cast

from neutron.nucleus._exec import Executor, require_nucleus
from neutron.nucleus.client import Features


class DatalogModel:
    """Datalog (logic programming) operations over Nucleus.

    Usage::

        await db.datalog.assert_fact("parent(alice, bob)")
        await db.datalog.rule("ancestor(X, Z)", "parent(X, Y), ancestor(Y, Z)")
        results = await db.datalog.query("ancestor(alice, ?X)")
    """

    def __init__(self, executor: Executor, features: Features) -> None:
        self._exec = executor
        self._features = features

    def _require(self) -> None:
        require_nucleus(self._features, "Datalog")

    async def assert_fact(self, fact: str) -> str:
        """Assert a Datalog fact. Returns the engine's status message."""
        self._require()
        return cast("str", await self._exec.fetchval("SELECT DATALOG_ASSERT($1)", fact))

    async def retract(self, fact: str) -> str:
        """Retract (remove) a previously asserted fact. Returns the engine's status message."""
        self._require()
        return cast("str", await self._exec.fetchval("SELECT DATALOG_RETRACT($1)", fact))

    async def rule(self, head: str, body: str) -> str:
        """Add a Datalog inference rule: ``head :- body``.

        The engine takes the whole rule as one string; head and body are
        joined here. Returns the engine's status message.
        """
        self._require()
        return cast(
            "str",
            await self._exec.fetchval("SELECT DATALOG_RULE($1)", f"{head} :- {body}")
        )

    async def query(self, query: str) -> list[list[str]]:
        """Execute a Datalog query.

        The engine returns a JSON array of arrays, e.g. ``[["alice","bob"]]``.
        """
        self._require()
        raw = await self._exec.fetchval("SELECT DATALOG_QUERY($1)", query)
        if not raw:
            return []
        return cast("list[list[str]]", json.loads(raw))

    async def clear(self, predicate: str) -> str:
        """Clear all facts and rules for a predicate. Returns the engine's status message."""
        self._require()
        return cast("str", await self._exec.fetchval("SELECT DATALOG_CLEAR($1)", predicate))

    async def import_graph(self, predicate: str) -> str:
        """Import graph edges as facts: ``predicate(from_id, edge_type, to_id)``.

        Returns the engine's status message (``IMPORTED N edges into <predicate>``).
        """
        self._require()
        return cast("str", await self._exec.fetchval("SELECT DATALOG_IMPORT_GRAPH($1)", predicate))

File: rust/crates/neutron-nucleus/src/models/datalog.rs
//! Datalog reasoning model — DATALOG_ASSERT, DATALOG_RETRACT, DATALOG_RULE,
//! DATALOG_QUERY, DATALOG_CLEAR, DATALOG_IMPORT_GRAPH.

use crate::error::NucleusError;
use crate::pool::NucleusPool;

/// Handle for Datalog reasoning operations.
pub struct DatalogModel {
    pool: NucleusPool,
}

impl DatalogModel {
    pub(crate) fn new(pool: NucleusPool) -> Self {
        Self { pool }
    }

    /// Assert a fact into the Datalog knowledge base.
    pub async fn assert_fact(&self, fact: &str) -> Result<bool, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_ASSERT($1)", &[&fact])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, bool>(0))
    }

    /// Retract a fact from the Datalog knowledge base.
    pub async fn retract(&self, fact: &str) -> Result<bool, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_RETRACT($1)", &[&fact])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, bool>(0))
    }

    /// Define a Datalog rule with a head and body.
    pub async fn rule(&self, head: &str, body: &str) -> Result<bool, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_RULE($1, $2)", &[&head, &body])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, bool>(0))
    }

    /// Evaluate a Datalog query. Returns results as CSV text.
    pub async fn query(&self, pattern: &str) -> Result<String, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_QUERY($1)", &[&pattern])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, String>(0))
    }

    /// Clear all facts and rules from the Datalog knowledge base.
    pub async fn clear(&self) -> Result<bool, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_CLEAR()", &[])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, bool>(0))
    }

    /// Import graph data into the Datalog knowledge base.
    /// Returns the number of facts imported.
    pub async fn import_graph(&self) -> Result<i64, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT DATALOG_IMPORT_GRAPH()", &[])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, i64>(0))
    }
}
